type.go

```diff
@@ -0,0 +1,458 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+// Package httprequest provides functionality for unmarshaling
+// http request parameters into a struct type.
+package httprequest
+
+import (
+	"encoding"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/julienschmidt/httprouter"
+	"gopkg.in/errgo.v1"
+)
+
+// TODO include field name and source in error messages.
+
+var (
+	typeMutex sync.RWMutex
+	typeMap   = make(map[reflect.Type]*requestType)
+)
+
+type Params struct {
+	*http.Request
+	Path httprouter.Params
+}
+
+// resultMaker is provided to the unmarshal functions.
+// When called with the value passed to the unmarshaler,
+// it returns the field value to be assigned to,
+// creating it if necessary.
+type resultMaker func(reflect.Value) reflect.Value
+
+// unmarshaler unmarshals some value from params into
+// the given value. The value should not be assigned to directly,
+// but passed to makeResult and then updated.
+type unmarshaler func(v reflect.Value, p Params, makeResult resultMaker) error
+
+// requestType holds information derived from a request
+// type, preprocessed so that it's nice and quick to unmarshal.
+type requestType struct {
+	fields []field
+}
+
+// field holds preprocessed information on an individual field
+// in the result.
+type field struct {
+	// index holds the index slice of the field.
+	index []int
+
+	// unmarshal is used to unmarshal the value into
+	// the given field. The value passed as its first
+	// argument is not a pointer type, but is addressable.
+	unmarshal unmarshaler
+
+	// makeResult is the resultMaker that will be
+	// passed into the unmarshaler.
+	makeResult resultMaker
+}
+
+// Unmarshal takes values from the request and URL parameters and fills
+// out fields in x, which must be a pointer to a struct.
+//
+// Tags on the struct's fields determine where each field is filled in
+// from. Similar to encoding/json and other encoding packages, the tag
+// holds a comma-separated list. The first item in the list is an
+// alternative name for the field (the field name itself will be used if
+// this is empty). The next item specifies where the field is filled in
+// from. It may be:
+//
+//	"url" - the field is taken from a parameter in urlParams
+//		with a matching field name.
+//
+// 	"form" - the field is taken the given name in req.Form
+//		(note that this covers both URL query parameters and
+//		POST form parameters)
+//
+//	"body" - the field is filled in by parsing the request body
```

Comment:
We should consider allowing this to be extended beyond json if necessary. I don't think anything ties us to json.